metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,911 @@
+            return (
+                isinstance(instance, PowerBiAPI.User)
+                and self.__members() == instance.__members()
+            )
+
+        def __hash__(self):
+            return hash(self.__members())
+
+    @dataclass
+    class Report:
+        id: str
+        name: str
+        webUrl: str
+        embedUrl: str
+        description: str
+        dataset: Optional["PowerBiAPI.PowerBIDataset"]
+        pages: List["PowerBiAPI.Page"]
+        users: List["PowerBiAPI.User"]
+
+        def get_urn_part(self):
+            return f"reports.{self.id}"
+
+    @dataclass
+    class Tile:
+        class CreatedFrom(Enum):
+            REPORT = "Report"
+            DATASET = "Dataset"
+            VISUALIZATION = "Visualization"
+            UNKNOWN = "UNKNOWN"
+
+        id: str
+        title: str
+        embedUrl: str
+        dataset: Optional["PowerBiAPI.PowerBIDataset"]
+        report: Optional[Any]
+        createdFrom: CreatedFrom
+
+        def get_urn_part(self):
+            return f"charts.{self.id}"
+
+    @dataclass
+    class Dashboard:
+        id: str
+        displayName: str
+        embedUrl: str
+        webUrl: str
+        isReadOnly: Any
+        workspace_id: str
+        workspace_name: str
+        tiles: List["PowerBiAPI.Tile"]
+        users: List["PowerBiAPI.User"]
+
+        def get_urn_part(self):
+            return f"dashboards.{self.id}"
+
+        def __members(self):
+            return (self.id,)
+
+        def __eq__(self, instance):
+            return (
+                isinstance(instance, PowerBiAPI.Dashboard)
+                and self.__members() == instance.__members()
+            )
+
+        def __hash__(self):
+            return hash(self.__members())
+
+    def __init__(self, config: PowerBiAPIConfig) -> None:
+        self.__config: PowerBiAPIConfig = config
+        self.__access_token: str = ""
+        # Power-Bi Auth (Service Principal Auth)
+        self.__msal_client = msal.ConfidentialClientApplication(
+            self.__config.client_id,
+            client_credential=self.__config.client_secret,
+            authority=PowerBiAPI.AUTHORITY + self.__config.tenant_id,
+        )
+
+        # Test connection by generating a access token
+        LOGGER.info("Trying to connect to {}".format(self.__get_authority_url()))
+        self.get_access_token()
+        LOGGER.info("Able to connect to {}".format(self.__get_authority_url()))
+
+    def __get_authority_url(self):
+        return "{}{}".format(PowerBiAPI.AUTHORITY, self.__config.tenant_id)
+
+    def __get_users(self, workspace_id: str, entity: str, _id: str) -> List[User]:
+        """
+        Get user for the given PowerBi entity
+        """
+        users: List[PowerBiAPI.User] = []
+        if self.__config.extract_ownership is False:
+            LOGGER.info(
+                "ExtractOwnership capabilities is disabled from configuration and hence returning empty users list"
```

Comment:
Extracting ownership is disabled from configuration... and so on

Reply:
Done